### ISSUE TITLE
Good procedure for a new species?

### ISSUE THREAD

**Issue by apredeus** [Dfam-consortium/RepeatModeler#40]
Hello Robert, 

I have a task of repeat masking of a reasonably well-assembled genome (N50 ~ 4 Mb, about 600 Mb in size). There are no related well-annotated species. As far as I understand, the correct approach is to first model repeats with RepeatModeler, and then mask them with RepeatMasker. However, I don't understand if I should use the database generated by Modeler alone, or a combination of default Dfam database + newly created DB? 

Another question: I've found this recommendation in one of the genomics blogs. Is that true? Should I only use big contigs (e.g. >1Mb in size) to run RepeatModeler? 

> My colleague .. pointed out that RepModeler does not seem to use all scaffolds/contigs to build a library, it samples them, so you might get slightly different results every time you run, and might be worth filtering out smaller contigs (e.g. <5000bp) so that it does not sample only small contigs for building the libraries

**Comment by apredeus** [comment]
So, to follow up. I've tried two approaches: 1) to model repeats from the whole assembly; 2) to model repeats only from contigs of 1Mb or larger. After this, I've used the obtained libraries to mask the whole genome assembly. Here's the output from approach 1: 

![image](https://user-images.githubusercontent.com/7825825/61959914-5538a980-afbc-11e9-92ab-df75d557d0d5.png)

And here's the output from approach 2: 

![image](https://user-images.githubusercontent.com/7825825/61959860-39350800-afbc-11e9-9b38-7c34e8659cbb.png)

So, quite strikingly, repeat classification is quite a bit better (11.9 vs 17.6% unclassified). Also, a lot more satellite DNA classified for the "full" library - I guess there's a lot more satellite repeats in small contigs, which would make sense. 

In the end, it's probably good to combine the two libraries using cd-hit and do a combo run?

**Comment by rmhubley** [comment]
I am finally catching up on requests -- sorry for the delay.  

These are excellent questions.  Ideally, one would use the RepeatModeler library as a starting point.  Curating each sequence it produced by extending fragments, building subfamilies, fixing classification, removing gene families and reconciling older families with previously identified TEs in related species.  This last step requires RepBase ( now subscription only ) or Dfam.   As these steps are quite involved, and until we can automate aspects of it we don't expect most users will do this.  Instead, we recommend that the the output of RepeatModeler combined with already defined repeats ( RepeatMasker/util/queryRepeatDatabase -species ... ) be placed into a custom library and then run through RepeatMasker with the "-lib" option.  

You answered your own question with the assembly preprocessing.  The current development version of RepeatModeler does provide more feedback to users about the characteristics of their input sequences:

```
  - Sequences = 790
  - Bases = 163915919
  - N50 = 7828983
  - Contig Histogram:
  Size(bp)                                                        Count
  -----------------------------------------------------------------------
  24034250-25750405 |                                                   [ 1 ]
  22318095-24034249 |                                                   [  ]
  20601940-22318094 |                                                   [ 1 ]
  18885785-20601939 |                                                   [  ]
  17169630-18885784 |                                                   [  ]
  15453475-17169629 |                                                   [  ]
  13737320-15453474 |                                                   [  ]
  12021166-13737320 |                                                   [ 1 ]
  10305011-12021165 |                                                   [  ]
  8588856-10305010  |                                                   [  ]
  6872701-8588855   |                                                   [ 1 ]
  5156546-6872700   |                                                   [  ]
  3440391-5156545   |                                                   [  ]
  1724236-3440390   |                                                   [ 5 ]
  8082-1724236      |************************************************** [ 780 ]
```

But, we need to go a step further and give users the option of biasing the sampling process.  An obvious one would be to bias towards larger contigs but allow at least a subset of smaller contigs to be in the mix -- for the reasons you point out.   Thanks for posting your classification results!

Finally, in response to the cd-hit question.  While, cd-hit ( or basic sequence clustering ) is helpful to find near perfect overlaps it is not a good measure of true genetic sequence distance ( e.g transition/transversion difference, affine gap penalties etc ).   We chose a sampling approach in RepeatModeler for two reasons.  The first being that the underlying de-novo codes ( RECON, RepeatScout ) were easily swamped by too much data.  Secondly, the iterative masking/discovery approach allowed us to avoid rediscovering the same families and avoid having to cluster at the end.  The only reason we didn't sample until 100% genome coverage had to do with compute power and overall runtime.  Since this is a single node, multi-threaded code it was hard to justify more than a week-long run.  Things have changed since we wrote this and now I would suggest adding on more rounds rather than running more than one independent analysis.  This is something we are working on in the next version -- due out in October.  

Let me know if you have any further questions.
